Implement reverse lookup (Ptr->Tuple) for CUDNN descriptors.

RomeoV opened the pull request:
This fixes #1947.

RomeoV committed:
Reverse lookup (Ptr->Tuple) for cudnn descriptors.

RomeoV commented:
Here's a small script that shows the loading and saving
```julia
import Pkg;
Pkg.activate("@CUDAPerfCachingTest")
# Setup:
# I don't think you can link to the cuDNN.jl module within CUDA.jl directly, so
# you'll have to clone github.com/romeov/CUDA.jl and then link
# Pkg.develop(path="<local>/romeov/CUDA.jl/lib/cudnn")
# Also
# Pkg.add("Flux")
# Pkg.add("JLD2")
#
# Execute e.g. with `julia caching_test.jl save` or `julia caching_test.jl load` or just `julia caching_test.jl`
using Flux, JLD2
import Flux.Zygote: gradient

function load_conv_caches!(; cudnn_mod::Module=Flux.cuDNN, filename="/tmp/conv_cache.jld2")
    @info "Loading conv_cache."
    conv_data_cache = JLD2.load(filename, "conv_data_cache");
    push!(cudnn_mod.cudnnConvolutionBwdDataAlgoPerfCache,
          conv_data_cache...)
    conv_filter_cache = JLD2.load(filename, "conv_filter_cache");
    push!(cudnn_mod.cudnnConvolutionBwdFilterAlgoPerfCache,
          conv_filter_cache...)
end

function save_conv_caches(; cudnn_mod::Module=Flux.cuDNN, filename="/tmp/conv_cache.jld2")
    @info "Storing conv_cache."
    JLD2.save(filename,
              "conv_data_cache", cudnn_mod.cudnnConvolutionBwdDataAlgoPerfCache,
              "conv_filter_cache", cudnn_mod.cudnnConvolutionBwdFilterAlgoPerfCache,
    )
end



if "load" in ARGS
    load_conv_caches!()
end

model = Chain(Conv((3, 3), 3=>64, relu; pad=SamePad()),
              Conv((3, 3), 64=>32, relu),
              GlobalMeanPool(),
              Flux.flatten,
              Dense(32=>1))

x = rand(Float32, 32, 32, 3, 7);

let x = gpu(x),
    model = gpu(model),
    ps = Flux.params(model)

  t0 = time()
  gradient(ps) do
      model(x) |> sum
  end
  @info "done in $(time() - t0) seconds :)"
end;
("load" in ARGS) && @show length(Flux.cuDNN.cudnnConvolutionBwdDataAlgoPerfCache)

if "save" in ARGS
    save_conv_caches()
end

```

ToucheSir commented on lib/cudnn/src/convolution.jl:
```diff
@@ -169,6 +169,12 @@ end
+# so that we can cache algorithms based on data descriptors.
+# Actually just reverses the cache dict and returns the descriptor as a tuple.
+map_cudnn_ptr_to_jl_tuple(cache_dict, desc_ptr) = Dict(zip(values(cache_dict),
+                                                           keys(cache_dict)))[desc_ptr]
```
Instead of recreating the cache in reversed form and searching it every time (expensive!), CUDA.jl provides functions for pulling out the info from a descriptor (cheap!). See https://github.com/JuliaGPU/CUDA.jl/blob/864ec5ee76a623227a73a762b5b6cca048224fd0/lib/cudnn/src/tensor.jl#L49-L72. You'll have to write equivalent functions for some of the conv-specific descriptors, but it should be quite straightforward.

RomeoV replied:
Done.

RomeoV replied:
The descriptors as they are still have some `Cenum` types in them, which we could convert to julia Ints or something if we run into serialization trouble.

RomeoV committed:
Add `cudnnGetConvolutionDescriptor` function

There is already `cudnnGetTensorDescriptor` and
`cudnnGetFilterDescriptor`, so now we have everything to cache algorithm performances.

RomeoV committed:
Store descriptor keys as native Julia dtypes

However, there's still a few `CUDNN_xyz_t` datatypes, which are Cenums.
We could still map those to Julia integers if serialization is difficult otherwise.

maleadt commented on lib/cudnn/src/convolution.jl:
```diff
@@ -179,11 +201,19 @@ const cudnnConvolutionFwdAlgoPerfCacheLock = ReentrantLock()
+    key = (xDesc_native, wDesc_native, convDesc_native, biasDesc, activation)
     val = lock(cudnnConvolutionFwdAlgoPerfCacheLock) do
-         get(cudnnConvolutionFwdAlgoPerfCache, key, nothing)
+        @debug "cudnnConvolutionFwdAlgoPerf: Checking if performance results are cached."
```
I wouldn't put debug messages on the hot path.

RomeoV replied:
Fair enough

RomeoV replied:
Removed.

maleadt commented on lib/cudnn/src/libcudnn.jl:
```diff
@@ -2927,51 +2927,51 @@ end
 end
 
 @checked function cudnnSetConvolution2dDescriptor(convDesc, pad_h, pad_w, u, v, dilation_h,
-                                                  dilation_w, mode, computeType)
```
These headers are auto-generated, so you shouldn't edit them.
Is the naming wrong maybe?

RomeoV replied:
Can you point me to where they were generated?
The naming was confusing to me when trying to understand how to use the API.
From my commit message:
> Not sure why it's called compute type here, but the C-side of the API
also just calls it `dataType` (see [here](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnGetConvolutionNdDescriptor)).

maleadt replied:
It's called like that in the CUDNN headers:

```
/usr/include/cudnn_cnn_infer.h
130:cudnnSetConvolution2dDescriptor(cudnnConvolutionDescriptor_t convDesc,
131-                                int pad_h,      /* zero-padding height */
132-                                int pad_w,      /* zero-padding width */
133-                                int u,          /* vertical filter stride */
134-                                int v,          /* horizontal filter stride */
135-                                int dilation_h, /* filter dilation in the vertical dimension */
136-                                int dilation_w, /* filter dilation in the horizontal dimension */
137-                                cudnnConvolutionMode_t mode,
138-                                cudnnDataType_t computeType);
```

The wrapper generator scripts are in `res/`, but I don't think it's worth carrying a patch or rewriting functionality to change a function argument name. So I'd suggest removing the change and leaving it be, even if it's a little unclear.

RomeoV replied:
Done.

RomeoV committed:
Remove debug messages.

maleadt commented:
This is still marked WIP; anything to do here @RomeoV @ToucheSir?

ToucheSir commented:
From my end no, didn't even notice the PR title still had WIP.

ToucheSir commented on lib/cudnn/src/convolution.jl:
```diff
@@ -241,7 +273,11 @@ const cudnnConvolutionBwdFilterAlgoPerfCacheLock = ReentrantLock()
+
+    key = (xDesc_native, dyDesc_native, convDesc_native)
     val = lock(cudnnConvolutionBwdFilterAlgoPerfCacheLock) do
         get(cudnnConvolutionBwdFilterAlgoPerfCache, (xDesc, dyDesc, convDesc), nothing)
```
@RomeoV whoops, I think I missed this line. It should be `get(cudnnConvolutionBwdFilterAlgoPerfCache, key, nothing)`, right?

RomeoV replied:
Good catch, thanks. Opened another PR with that one-line change.
